File: inference/js/infer.js
/**
 * Node.js Inference Runner for Bluestone Indic TTS.
 * Generates spoken audio and saves to WAV format directly.
 */

const fs = require('fs');
const path = require('path');
const ort = require('onnxruntime-node');
const { IndicTokenizer } = require('./indic_tokenizer');

function writeWavFile(filepath, samples, sampleRate = 22050) {
  // Convert float32 [-1.0, 1.0] to 16-bit PCM
  const buffer = Buffer.alloc(44 + samples.length * 2);

  // RIFF header
  buffer.write('RIFF', 0);
  buffer.writeUInt32LE(36 + samples.length * 2, 4);
  buffer.write('WAVE', 8);

  // fmt chunk
  buffer.write('fmt ', 12);
  buffer.writeUInt32LE(16, 16);          // Subchunk1Size (16 for PCM)
  buffer.writeUInt16LE(1, 20);           // AudioFormat (1 = PCM)
  buffer.writeUInt16LE(1, 22);           // NumChannels (1 = Mono)
  buffer.writeUInt32LE(sampleRate, 24);  // SampleRate
  buffer.writeUInt32LE(sampleRate * 2, 28); // ByteRate (SampleRate * NumChannels * BitsPerSample/8)
  buffer.writeUInt16LE(2, 32);           // BlockAlign (NumChannels * BitsPerSample/8)
  buffer.writeUInt16LE(16, 34);          // BitsPerSample (16 bits)

  // data chunk
  buffer.write('data', 36);
  buffer.writeUInt32LE(samples.length * 2, 40);

  let offset = 44;
  for (let i = 0; i < samples.length; i++) {
    let s = Math.max(-1.0, Math.min(1.0, samples[i]));
    let val = s < 0 ? s * 32768 : s * 32767;
    buffer.writeInt16LE(Math.floor(val), offset);
    offset += 2;
  }

  fs.writeFileSync(filepath, buffer);
}

async function runTTS(modelPath, vocabPath, text, outputPath = 'output_node.wav', speed = 1.0) {
  console.log(`[*] Loading ONNX model: ${modelPath}`);
  const tokenizer = new IndicTokenizer(vocabPath);

  const session = await ort.InferenceSession.create(modelPath, {
    executionProviders: ['cpu'],
    graphOptimizationLevel: 'all'
  });

  const tokens = tokenizer.encode(text, true, true);
  console.log(`[*] Encoded ${tokens.length} phonetic tokens`);

  const lengthScale = 1.0 / Math.max(0.2, speed);
  const inputIdsTensor = new ort.Tensor('int64', BigInt64Array.from(tokens.map(t => BigInt(t))), [1, tokens.length]);
  const inputLengthsTensor = new ort.Tensor('int64', BigInt64Array.from([BigInt(tokens.length)]), [1]);
  const scalesTensor = new ort.Tensor('float32', Float32Array.from([0.667, lengthScale, 0.8]), [3]);

  const feeds = {
    input_ids: inputIdsTensor,
    input_lengths: inputLengthsTensor,
    scales: scalesTensor
  };

  const startTime = Date.now();
  const results = await session.run(feeds);
  const latencyMs = Date.now() - startTime;

  const audioTensor = results.audio;
  const audioData = audioTensor.data; // Float32Array
  const sampleRate = 22050;
  const durationSec = audioData.length / sampleRate;
  const rtf = (latencyMs / 1000) / durationSec;

  writeWavFile(outputPath, audioData, sampleRate);

  console.log(`[✓] Audio generated successfully!`);
  console.log(`    Output:      ${outputPath}`);
  console.log(`    Duration:    ${durationSec.toFixed(2)}s`);
  console.log(`    Latency:     ${latencyMs}ms`);
  console.log(`    RTF on CPU:  ${rtf.toFixed(3)} (${(1 / rtf).toFixed(1)}x real-time speed)`);
}

// Example execution
if (require.main === module) {
  const modelPath = process.argv[2] || path.join(__dirname, '../../exported/bluestone_tts_hi.onnx');
  const vocabPath = process.argv[3] || path.join(__dirname, '../../frontend/vocab.json');
  const text = process.argv[4] || 'नमस्ते! यह नोड जेएस में चल रहा ब्लूस्टोन टेक्स्ट टू स्पीच इंजन है।';
  const outPath = process.argv[5] || 'output_node.wav';

  runTTS(modelPath, vocabPath, text, outPath).catch(err => {
    console.error('[!] Error running TTS in Node.js:', err);
  });
}

module.exports = { runTTS };

File: inference/python/infer.py
"""
Python ONNX Runtime Inference Engine.
Synthesizes speech from text using exported Bluestone TTS ONNX models.
"""

import os
import argparse
import numpy as np
import onnxruntime as ort
import soundfile as sf
from frontend.indic_tokenizer import IndicTokenizer

class BluestoneTTS:
    def __init__(self, model_path: str, vocab_path: str = None, num_threads: int = 4):
        self.model_path = model_path
        if not os.path.exists(model_path):
            raise FileNotFoundError(f"Model file not found: {model_path}")

        sess_options = ort.SessionOptions()
        sess_options.graph_optimization_level = ort.GraphOptimizationLevel.ORT_ENABLE_ALL
        sess_options.intra_op_num_threads = num_threads

        self.session = ort.InferenceSession(model_path, sess_options, providers=["CPUExecutionProvider"])
        
        # Read model metadata
        meta = self.session.get_modelmeta().custom_metadata_map
        self.sample_rate = int(meta.get("sample_rate", 22050))
        self.lang = meta.get("language", "hi")

        self.tokenizer = IndicTokenizer(vocab_path=vocab_path, lang=self.lang)

    def synthesize(
        self,
        text: str,
        speed: float = 1.0,
        noise_scale: float = 0.667,
        noise_scale_w: float = 0.8
    ) -> np.ndarray:
        """
        Synthesize speech from Indic text.
        Args:
            text: Input text in Indic script (e.g. Hindi Devanagari)
            speed: Speaking speed multiplier (1.0 = normal, 1.2 = faster, 0.8 = slower)
            noise_scale: Phoneme variation noise scale
            noise_scale_w: Duration variation noise scale
        Returns:
            audio_array: 1D numpy float32 array
        """
        # Convert speed to length_scale (speed 1.25 -> length_scale 0.8)
        length_scale = 1.0 / max(0.2, speed)

        tokens = self.tokenizer.encode(text, add_bos_eos=True, add_blank=True)
        input_ids = np.array([tokens], dtype=np.int64)
        input_lengths = np.array([len(tokens)], dtype=np.int64)
        scales = np.array([noise_scale, length_scale, noise_scale_w], dtype=np.float32)

        inputs = {
            "input_ids": input_ids,
            "input_lengths": input_lengths,
            "scales": scales
        }

        outputs = self.session.run(None, inputs)
        audio = outputs[0]  # [1, 1, num_samples]
        audio = audio.squeeze()
        return audio

    def save_wav(self, audio: np.ndarray, output_path: str):
        """Save synthesized audio array to WAV file."""
        os.makedirs(os.path.dirname(os.path.abspath(output_path)), exist_ok=True)
        # Normalize and prevent clipping
        max_val = np.max(np.abs(audio))
        if max_val > 1.0:
            audio = audio / max_val * 0.99
        sf.write(output_path, audio, self.sample_rate)
        print(f"[✓] Audio saved to: {output_path} ({len(audio)/self.sample_rate:.2f}s)")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Run Bluestone TTS inference")
    parser.add_argument("--model", type=str, default="exported/bluestone_tts_hi.onnx", help="Path to ONNX model")
    parser.add_argument("--vocab", type=str, default="frontend/vocab.json", help="Path to vocab.json")
    parser.add_argument("--text", type=str, default="नमस्ते, यह ब्लूस्टोन टेक्स्ट टू स्पीच का परीक्षण है।", help="Text to speak")
    parser.add_argument("--output", type=str, default="output.wav", help="Path to output WAV file")
    parser.add_argument("--speed", type=float, default=1.0, help="Speaking speed (e.g. 1.0, 1.2)")
    args = parser.parse_args()

    engine = BluestoneTTS(model_path=args.model, vocab_path=args.vocab)
    audio = engine.synthesize(args.text, speed=args.speed)
    engine.save_wav(audio, args.output)
